### PR TITLE
subscription poc

### PR DESCRIPTION
**Pull request by dhiraka**
I want to merge this change because...

# Impact

* [x] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations

# Pull Request Checklist

* [x] Privileged queries and mutations are guarded by proper permission checks
* [x] Database queries are optimized and the number of queries is constant
* [x] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [x] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by pankaj28843** [comment]
https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716

https://help.github.com/en/github/committing-changes-to-your-project/changing-a-commit-message

File: saleor/graphql/subscription/mutations.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        created = instance.pk
+        super().save(info, instance, cleaned_input)
+
+        # # Create draft created event if the instance is from scratch
+        # if not created:
+        #     events.draft_subscription_created_event(
+        #         subscription=instance,
+        #         user=info.context.user
+        #     )
+
```
Comment (pankaj28843) [inline saleor/graphql/subscription/mutations.py]: Please delete unused code

File: saleor/graphql/subscription/schema.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import graphene
+
+from .resolvers import resolve_subscription, resolve_subscriptions, \
+    resolve_subscription_by_token, resolve_draft_subscriptions
+from ...core.permissions import SubscriptionPermissions
+from ..core.fields import FilterInputConnectionField, PrefetchingConnectionField
+from ..decorators import permission_required
+from .mutations import (
```
Comment (pankaj28843) [inline saleor/graphql/subscription/schema.py:3]: Please sort imports using IDE
https://www.jetbrains.com/pycharm/guide/tips/optimize-imports/

File: saleor/graphql/subscription/schema.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+class SubscriptionQueries(graphene.ObjectType):
+    subscription = graphene.Field(
+        Subscription,
+        description="Look up a subscription by ID.",
+        id=graphene.Argument(graphene.ID, description="ID of a subscription.", required=True),
+    )
+    subscriptions = FilterInputConnectionField(
+        Subscription,
+#         sort_by=SubscriptionSortingInput(description="Sort subscriptions."),
```
Comment (pankaj28843) [inline saleor/graphql/subscription/schema.py]: Please remove this

File: saleor/graphql/subscription/mutations.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        user = info.context.user
+        if not user.is_authenticated:
+            raise PermissionDenied()
+        subscription = cls.get_instance(info, **data)
+        cleaned_input = cls.clean_input(info, subscription, data.get("input"))
+        cleaned_input['status'] = SubscriptionStatus.ENDED
+        subscription = cls.construct_instance(subscription, cleaned_input)
+        cls.save(info, subscription, cleaned_input)
```
Comment (pankaj28843) [inline saleor/graphql/subscription/mutations.py]: Let's create methods on Subscription model to take action on subscription objects. Then, the logic for actions can be reused when we write background tasks or other python scripts

**Review by pankaj28843 — CHANGES_REQUESTED** [review]

File: saleor/subscription/models.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    def save(self, *args, **kwargs):
+        if not self.token:
+            self.token = str(uuid4())
+        return super().save(*args, **kwargs)
+
+    def end(self, *args, **kwargs):
+        self.status=SubscriptionStatus.ENDED
+        if 'ended_with_reason' in kwargs:
+            self.ended_with_reason=kwargs["ended_with_reason"]
```
Comment (pankaj28843) [inline saleor/subscription/models.py]: space around ` = `;

File: saleor/subscription/models.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+    def get_customer_email(self):
+        return self.user.email if self.user else self.user_email
```
Comment (pankaj28843) [inline saleor/subscription/models.py]: `user_email` field does not exist in this model. Guest users won't be allowed to subscribe.

**Review by pankaj28843 — APPROVED** [review]

File: saleor/subscription/models.py (hunk 6 above)
Comment (dhiraka) [reply saleor/subscription/models.py]: Guest users should not be able to subscribe rt?

Comment (dhiraka) [reply saleor/subscription/models.py]: removing the method

**Review by pankaj28843 — APPROVED** [review]

**Pull request merged**